simSPI/linear_simulator/modules.py

```diff
@@ -0,0 +1,397 @@
+        -------
+        projection: tensor
+            Tensor containing tomographic projection
+            (chunks x 1 x sidelen x sidelen)
+        """
+        rotmat = rot_params["rotmat"]
+        batch_sz = rotmat.shape[0]
+        t = Rotate(rotmat, device=self.vol_coords.device)
+        rot_vol_coords = t.transform_points(self.vol_coords.repeat(batch_sz, 1, 1))
+
+        rot_vol = torch.nn.functional.grid_sample(
+            self.vol.repeat(batch_sz, 1, 1, 1, 1),  # = (Batch, C,D,H,W)
+            rot_vol_coords[:, None, None, :, :],
+            align_corners=True,
+        )
+        projection = torch.sum(
+            rot_vol.reshape(
+                batch_sz, self.config.sidelen, self.config.sidelen, self.config.sidelen
+            ),
+            dim=proj_axis,
+        )
+        projection = projection[
+            :, None, :, :
+        ]  # add a dummy channel (for consistency w/ img fmt)
+        # --> B,C,H,W
+        return projection
+
+
+"""Module that multiplies ctf to fourier transform of the projection."""
+
+
+class CTF(torch.nn.Module):
+    """Class that multiplies ctf to fourier transform of the projection.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the CTF
+
+    """
+
+    def __init__(self, config):
+        super(CTF, self).__init__()
+
+        self.config = config
+        self.wavelength = self._get_ewavelength()
+        self.frequency = 1.0 / (self.config.ctf_size * self.config.pixel_size)
+        n2 = float(self.config.ctf_size // 2)
+        ax = torch.arange(-n2, n2 + self.config.ctf_size % 2)
+        mx, my = torch.meshgrid(ax, ax)
+        self.register_buffer("r2", mx ** 2 + my ** 2)
+        self.register_buffer("r", torch.sqrt(self.r2))
+        self.register_buffer("angleFrequency", torch.atan2(my, mx))
+
+    def _get_ewavelength(self):
+        """Output the wavelength of the electron beam.
+
+        Returns
+        -------
+        wavelength: float
+            wavelength of the electron beam
+        """
+        wavelength = 12.2639 / np.sqrt(
+            self.config.kV * 1e3 + 0.97845 * self.config.kV ** 2
+        )
+        return wavelength
+
+    def get_ctf(self, ctf_params):
+        """Output the CTF for a given params.
+
+        Parameters
+        ----------
+        ctf_params: dict of type str to {tensor}
+            contains "defocusU", "defocusV", "defocusAngle",
+            all of shape (chunks,1,1,1)
+
+        Returns
+        -------
+        hFourier: torch.Tensor
+            ctf of shape (chunks,1,ctf_size,ctf_size)
+        """
+        defocusU = ctf_params["defocusU"]
+        defocusV = ctf_params["defocusV"]
+        angleAstigmatism = ctf_params["defocusAngle"]
+        elliptical = (
+            defocusV * self.r2
+            + (defocusU - defocusV)
+            * self.r2
+            * torch.cos(self.angleFrequency - angleAstigmatism) ** 2
+        )
+        defocusContribution = (
```

Comment:
math looks good. at least up to that factor of 1e4 in line 217.

if you had units in the variable names, it would make it easier to verify.